### DISCUSSION TITLE
Error When Trying to Deploy Contract Using Infura and Rinkeby

### DISCUSSION THREAD

**Discussion by maxdporter** [smartcontractkit/full-blockchain-solidity-course-py#811] [Q&A] [answered]
**I was able to deploy my contract to ganache without issue in my terminal. 

However, when I update the rinkeby.infura endpoint and all realted fields like below:**

w3 = Web3(
    Web3.HTTPProvider("https://rinkeby.infura.io/v3/2dd0a8bc2aa34d72b48e111ecf473caf")
)
chain_id = 4
my_address = "0x20e73C456cF9A3fE6F80E112A5cAC077c0A8728D"
private_key = os.getenv("PRIVATE_KEY")

**I receive this error and am not sure how to correct it:**
Installing...
Traceback (most recent call last):
  File "/Users/maxporter/demos/web3_py_simple_storage/deploy.py", line 66, in <module>
    transaction = SimpleStorage.constructor().buildTransaction(
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/eth_utils/decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/contract.py", line 684, in buildTransaction
    return fill_transaction_defaults(self.web3, built_transaction)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/_utils/transactions.py", line 114, in fill_transaction_defaults
    default_val = default_getter(web3, transaction)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/_utils/transactions.py", line 60, in <lambda>
    'gas': lambda web3, tx: web3.eth.estimate_gas(tx),
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/eth.py", line 820, in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/manager.py", line 198, in request_blocking
    return self.formatted_response(response,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/web3/manager.py", line 171, in formatted_response
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'gas required exceeds allowance (0)'}

**Comment by cromewar** [comment, marked as answer]
Hello @maxdporter there are some reasons this might happen.

1. first of all please share your entire code here (deploy.py and contract)
2. second I see some parts of your error refer to Cytools which is a pretty common error on this field, you would like to install visual studio community and it's python development tools or use WSL instead.
3. python 3.10.1 has been reported to have problems with web3 libraries, using python 3.8 might help.

**Reply by maxdporter** [reply]
**@cromewar I appreciate your help, here's the deploy contract**

```Python
import json

from web3 import Web3

# In the video, we forget to `install_solc`
# from solcx import compile_standard
from solcx import compile_standard, install_solc
import os
from dotenv import load_dotenv

load_dotenv()


with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

# We add these two lines that we forgot from the video!
print("Installing...")
install_solc("0.6.0")

# Solidity source code
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                }
            }
        },
    },
    solc_version="0.6.0",
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

# get bytecode
bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

# get abi
abi = json.loads(
    compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["metadata"]
)["output"]["abi"]

# w3 = Web3(Web3.HTTPProvider(os.getenv("RINKEBY_RPC_URL")))
# chain_id = 4
#
# For connecting to ganache
w3 = Web3(
    Web3.HTTPProvider("https://rinkeby.infura.io/v3/2dd0a8bc2aa34d72b48e111ecf473caf")
)
chain_id = 4
my_address = "0x20e73C456cF9A3fE6F80E112A5cAC077c0A8728D"
private_key = os.getenv("PRIVATE_KEY")

# Create the contract in Python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)
# Get the latest transaction
nonce = w3.eth.getTransactionCount(my_address)
# Submit the transaction that deploys the contract
transaction = SimpleStorage.constructor().buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce,
    }
)
# Sign the transaction
signed_txn = w3.eth.account.sign_transaction(transaction, private_key=private_key)
print("Deploying Contract!")
# Send it!
tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
# Wait for the transaction to be mined, and get the transaction receipt
print("Waiting for transaction to finish...")
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
print(f"Done! Contract deployed to {tx_receipt.contractAddress}")


# Working with deployed Contracts
simple_storage = w3.eth.contract(address=tx_receipt.contractAddress, abi=abi)
print(f"Initial Stored Value {simple_storage.functions.retrieve().call()}")
greeting_transaction = simple_storage.functions.store(15).buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce + 1,
    }
)
signed_greeting_txn = w3.eth.account.sign_transaction(
    greeting_transaction, private_key=private_key
)
tx_greeting_hash = w3.eth.send_raw_transaction(signed_greeting_txn.rawTransaction)
print("Updating stored Value...")
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_greeting_hash)

print(simple_storage.functions.retrieve().call())
```

**and here is the solidity file**

```Solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.6.0;

contract SimpleStorage {
    // this will get initialized to 0!
    uint256 favoriteNumber;
    bool favoriteBool;

    struct People {
        uint256 favoriteNumber;
        string name;
    }

    People[] public people;
    mapping(string => uint256) public nameToFavoriteNumber;

    function store(uint256 _favoriteNumber) public {
        favoriteNumber = _favoriteNumber;
    }

    function retrieve() public view returns (uint256) {
        return favoriteNumber;
    }

    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }
}

```

**Reply by cromewar** [reply]
Seems everything is ok with your code, so please try with points 2 and 3 I mentioned above

**Comment by maxdporter** [comment]
For point 2, could you link to those tools?
For point 3, do you know how I could run a program with a specific version of python?

**Reply by maxdporter** [reply]
(I'm on mac os)

**Reply by cromewar** [reply]
As you are on Mac you should install xcode and cython.

Also yes, on the bottom left corner on vscode you can change between python interpreters, using python virtual envs or conda will change the version.

**Reply by maxdporter** [reply]
Thanks @cromewar, I installed cython, changed python version to 3.8.9, AND changed the install command from pipx to pip3. That seemed to do the trick

**Reply by cromewar** [reply]
Nice, I'm glad it helped you
